In [12]:
#importing libraries

import pandas as pd
from sklearn.tree import DecisionTreeClassifier 
from matplotlib import pyplot as plt
import numpy as np
from sklearn.metrics import confusion_matrix
from sklearn import preprocessing
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
from scipy import stats
np.random.seed(1)

In [13]:
X_train = pd.read_csv("C:/Users/ssand/OneDrive/Desktop/v1/airbnb_train_X_price_gte_150.csv")
X_test = pd.read_csv("C:/Users/ssand/OneDrive/Desktop/v1/airbnb_test_X_price_gte_150.csv")
y_train = pd.read_csv("C:/Users/ssand/OneDrive/Desktop/v1/airbnb_train_y_price_gte_150.csv")
y_test = pd.read_csv("C:/Users/ssand/OneDrive/Desktop/v1/airbnb_test_y_price_gte_150.csv")


In [14]:
#SVM classification using Random search

score_measure = "precision"
kfolds = 5

param_grid = {
     'C': [0.1,1, 10, 100], 
    'gamma': [1,0.1,0.01,0.001],
    'kernel': ['poly']
}

RN_search = SVC()
rand_search = RandomizedSearchCV(estimator = RN_search, param_distributions=param_grid, cv=kfolds, n_iter=500,
                           scoring=score_measure, verbose=1, n_jobs=-1,
                           return_train_score=True)

_ = rand_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {rand_search.best_score_}")
print(f"... with parameters: {rand_search.best_params_}")

bestPrecisionTree = rand_search.best_estimator_



C:\Users\ssand\AppData\Roaming\Python\Python39\site-packages\sklearn\model_selection\_search.py:306: UserWarning: The total space of parameters 16 is smaller than n_iter=500. Running 16 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(


Fitting 5 folds for each of 16 candidates, totalling 80 fits


C:\Users\ssand\AppData\Roaming\Python\Python39\site-packages\sklearn\utils\validation.py:1111: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


The best precision score is 0.9370404920282969
... with parameters: {'kernel': 'poly', 'gamma': 0.01, 'C': 0.1}


The best precision score is 0.9370404920282969

In [15]:
#SVM classification using Grid search


score_measure = "precision"
kfolds = 5

param_grid = {
     'C': [0.1,1, 10, 100], 
     'gamma': [1,0.1,0.01,0.001],
    'kernel': ['poly']
}

G_search = SVC()
grid_search = GridSearchCV(estimator = G_search, param_grid=param_grid, cv=kfolds, 
                           scoring=score_measure, verbose=1, n_jobs=-1,
                           return_train_score=True)

_ = grid_search.fit(X_train, y_train)

print(f"The perfect  {score_measure} score as far as  {grid_search.best_score_}")
print(f"... with parameters: {grid_search.best_params_}")

bestPrecisionTree = grid_search.best_estimator_

Fitting 5 folds for each of 16 candidates, totalling 80 fits


C:\Users\ssand\AppData\Roaming\Python\Python39\site-packages\sklearn\utils\validation.py:1111: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


The perfect  precision score as far as  0.9370404920282969
... with parameters: {'C': 0.1, 'gamma': 0.01, 'kernel': 'poly'}


The perfect  precision score as far as  0.9370404920282969

In [16]:
#Decision tree using Random search

score_measure = "precision"
kfolds = 5

param_grid = {
    'min_samples_split': np.arange(1,70),  
    'min_samples_leaf': np.arange(1,80),
    'min_impurity_decrease': np.arange(0.0001, 0.01, 0.0005),
    'max_leaf_nodes': np.arange(5, 100), 
    'max_depth': np.arange(1,50), 
    'criterion': ['entropy', 'gini'],
}

dtree = DecisionTreeClassifier()
rand_search = RandomizedSearchCV(estimator = dtree, param_distributions=param_grid, cv=kfolds, n_iter=500,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = rand_search.fit(X_train, y_train)

print(f"The  {score_measure} score is {rand_search.best_score_}")
print(f"... with parameters: {rand_search.best_params_}")

bestPrecisionTree = rand_search.best_estimator_

Fitting 5 folds for each of 500 candidates, totalling 2500 fits
The  precision score is 0.8570532608300236
... with parameters: {'min_samples_split': 21, 'min_samples_leaf': 5, 'min_impurity_decrease': 0.0016, 'max_leaf_nodes': 37, 'max_depth': 25, 'criterion': 'entropy'}


C:\Users\ssand\AppData\Roaming\Python\Python39\site-packages\sklearn\model_selection\_validation.py:378: FitFailedWarning: 
20 fits failed out of a total of 2500.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
20 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\ssand\AppData\Roaming\Python\Python39\site-packages\sklearn\model_selection\_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\ssand\AppData\Roaming\Python\Python39\site-packages\sklearn\tree\_classes.py", line 969, in fit
    super().fit(
  File "C:\Users\ssand\AppData\Roaming\Python\Python39\site-packages\sklearn\tree\_classes.py", line 265, in fit
    check_scalar(
  File "C:\

The  precision score is 0.8570532608300236

In [17]:
#Decision tree using Grid search
score_measure = "precision"
kfolds = 5

param_grid = {
    'min_samples_split': np.arange(18,24),  
    'min_samples_leaf': np.arange(3,7),
    'min_impurity_decrease': np.arange(0.0010, 0.0017, 0.0001),
    'max_leaf_nodes': np.arange(34,40), 
    'max_depth': np.arange(23,27), 
    'criterion': ['entropy'],
}

Dout_GS = DecisionTreeClassifier()
grid_search = GridSearchCV(estimator = Dout_GS, param_grid=param_grid, cv=kfolds, 
                           scoring=score_measure, verbose=1, n_jobs=-1,
                           return_train_score=True)

_ = grid_search.fit(X_train, y_train)

print(f"The {score_measure} score is {grid_search.best_score_}")
print(f"... with parameters: {grid_search.best_params_}")

bestPrecisionTree = grid_search.best_estimator_

Fitting 5 folds for each of 4032 candidates, totalling 20160 fits
The precision score is 0.8613399498900961
... with parameters: {'criterion': 'entropy', 'max_depth': 23, 'max_leaf_nodes': 35, 'min_impurity_decrease': 0.001, 'min_samples_leaf': 4, 'min_samples_split': 20}


Conclusion from above :
    
    Clearly, we can see that the Svm Classification is performming better compared to decision tree with the given data set. So SVM is the better performming model here.
    
    * The SVM with random search has percision value as 0.93 and for SVM with grid search has 0.93.
    
    * The Decision tree with random search has percision value is 0.85 and for the Decision tree with grid search has percision value as 0.86 if we compare in between both internally the decision tree with random search is performing better.
   
    